File: server/state.py
from dataclasses import dataclass
from threading import Lock


@dataclass
class Deployment:
    commit: str
    branch: str
    image_tag: str
    container_id: str
    url: str


class AppState:
    def __init__(self) -> None:
        self._lock = Lock()
        self._deployment: Deployment | None = None

    def current(self) -> Deployment | None:
        with self._lock:
            return self._deployment

    def swap(self, next: Deployment) -> Deployment | None:
        with self._lock:
            prev = self._deployment
            self._deployment = next
            return prev

File: server/src/state.rs
use std::sync::{Arc, Mutex};

use serde::Serialize;

#[derive(Clone, Serialize)]
pub struct Deployment {
    pub commit: String,
    pub branch: String,
    pub image_tag: String,
    pub container_id: String,
    pub url: String,
}

#[derive(Clone, Default)]
pub struct AppState {
    deployment: Arc<Mutex<Option<Deployment>>>,
}

impl AppState {
    pub fn current(&self) -> Option<Deployment> {
        self.deployment.lock().expect("state lock").clone()
    }

    pub fn swap(&self, next: Deployment) -> Option<Deployment> {
        self.deployment.lock().expect("state lock").replace(next)
    }
}
